src/wire/ipv6.rs

```diff
@@ -0,0 +1,372 @@
+    ///
+    /// # Panics
+    /// The function panics if `data` is not 8 words long
+    pub fn from_parts(data: &[u16]) -> Address {
+        let mut bytes = [0; 16];
+        for word_idx in 0..8 {
+            let byte_idx = word_idx * 2;
+            NetworkEndian::write_u16(&mut bytes[byte_idx..(byte_idx + 2)], data[word_idx]);
+        }
+        Address(bytes)
+    }
+
+    /// Write a IPv6 address to the given slice
+    ///
+    /// # Panics
+    /// The function panics if `data` is not 8 words long
+    pub fn write_parts(&self, data: &mut [u16]) {
+        for i in 0..8 {
+            let byte_idx = i * 2;
+            data[i] = NetworkEndian::read_u16(&self.0[byte_idx..(byte_idx + 2)]);
+        }
+    }
+
+    /// Return an IPv6 address as a sequence of octets, in big-endian.
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+
+    /// Query whether the IPv6 address is an unicast address.
+    pub fn is_unicast(&self) -> bool {
+        !(self.is_multicast() || self.is_unspecified())
+    }
+
+    /// Query whether the IPv6 address is a multicast address.
+    pub fn is_multicast(&self) -> bool {
+        self.0[0] == 0xff
+    }
+
+    /// Query whether the IPv6 address is the "unspecified" address
+    pub fn is_unspecified(&self) -> bool {
+        self.0 == [0x00; 16]
+    }
+
+    /// Query whether the IPv6 address is in the "link-local" range
+    pub fn is_link_local(&self) -> bool {
+        self.0[0..8] == [0xfe, 0x80, 0x00, 0x00,
+                         0x00, 0x00, 0x00, 0x00]
+    }
+
+    /// Query whether the IPv6 address is the "loopback" address
+    pub fn is_loopback(&self) -> bool {
+        self.0 == [0x00, 0x00, 0x00, 0x00,
+                   0x00, 0x00, 0x00, 0x00,
+                   0x00, 0x00, 0x00, 0x00,
+                   0x00, 0x00, 0x00, 0x01]
+    }
+
+    /// Helper function used to mask an addres given a prefix
+    ///
+    /// # Panics
+    /// This function panics if `mask` is greater than 128
+    pub(super) fn mask(&self, mask: u8) -> [u8; 16] {
+        assert!(mask <= 128);
+        let mut bytes = [0u8; 16];
+        let idx = (mask as usize) / 8;
+        let modulus = (mask as usize) % 8;
+        let (first, second) = self.0.split_at(idx);
+        bytes[0..idx].copy_from_slice(&first);
+        if idx < 16 {
+            let part = second[0];
+            bytes[idx] = part & (!(0xff >> modulus) as u8);
+        }
+        bytes
+    }
+}
+
+impl fmt::Display for Address {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let bytes = self.0;
+        write!(f,
+               "{:02x}{:02x}:{:02x}{:02x}:{:02x}{:02x}:{:02x}{:02x}:\
+               {:02x}{:02x}:{:02x}{:02x}:{:02x}{:02x}:{:02x}{:02x}",
+               bytes[0], bytes[1], bytes[2], bytes[3],
+               bytes[4], bytes[5], bytes[6], bytes[7],
+               bytes[8], bytes[9], bytes[10], bytes[11],
+               bytes[12], bytes[13], bytes[14], bytes[15])
```

Comment:
This could really use a very simple `::` shortener (start munging at the first `0u16` part).

Reply:
I took a very simple approach and added the IPv4 mapped IPv6 address case. Ipv4 Compatible addresses have been deprecated, so there is no need to add a case for them. I thought about essentially copy and pasting https://github.com/rust-lang/rust/blob/1.19.0/src/libstd/net/ip.rs#L1113-L1182, but slice pattern matching is used